src/plugins/cordova-plugin-device-motion/device-motion-model.js

```diff
@@ -1,47 +1,176 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 
+Math.sign = Math.sign || function (x) {
+    x = +x; // convert to a number
+    if (x === 0 || isNaN(x)) {
+        return Number(x);
+    }
+    return x > 0 ? 1 : -1;
+};
+
 // report interval in milliseconds
 var ACCELEROMETER_REPORT_INTERVAL = 50;
+var G_CONSTANT = 9.81;
+var degToRad = Math.PI / 180;
 
-var _onChangeCallback,
-    _axis = {
-        x: 0,
-        y: 0,
-        z: 0
-    };
+var _onChangeCallback;
+var axisValues = ['x', 'y', 'z'];
+var rotationValues = ['alpha', 'beta', 'gamma'];
+var allValues = axisValues.concat(rotationValues);
+
+var _axis = {
+    x: 0,
+    y: 0,
+    z: 0
+};
+normalizeFromAcceleration(_axis);
 
 function updateAxis(values) {
+    var newValues = applyValues(values, axisValues);
+    normalizeFromAcceleration(newValues);
+    update(newValues);
+}
+
+function updateRotation(values) {
+    var newValues = applyValues(values, rotationValues);
+    normalizeFromRotation(newValues);
+    update(newValues);
+}
+
+function addAxisChangeCallback(callback) {
+    _onChangeCallback = (typeof callback === 'function') ? callback : null;
+}
+
+function applyValues(values, names) {
+    var newValues = JSON.parse(JSON.stringify(_axis));
+    names.forEach(function (name) {
+        if (values.hasOwnProperty(name)) {
+            newValues[name] = values[name];
+        }
+    });
+    return newValues;
+}
+
+function update(values) {
     values = values || {};
 
     var updatedValues = {};
+    var dirty = false;
+
+    allValues.forEach(function (name) {
+        if (values[name] !== _axis[name]) {
+            dirty = true;
+            _axis[name] = values[name];
+            updatedValues[name] = values[name];
+        }
+    });
+
+    // notify if any value has been updated
+    if (dirty && _onChangeCallback) {
+        _onChangeCallback(updatedValues);
+    }
+}
+
+function normalizeFromRotation(values) {
+    values.alpha = constrainAlpha(values.alpha);
+    values.beta = constrainBeta(values.beta);
+    values.gamma = constrainGamma(values.gamma);
+
+    var accel = accelFromRot(values.beta, values.gamma);
+    values.x = accel.x;
+    values.y = accel.y;
+    values.z = accel.z;
+}
+
+function normalizeFromAcceleration(values) {
+    values.x = constrainAcceleration(values.x);
+    values.y = constrainAcceleration(values.y);
+    values.z = constrainAcceleration(values.z);
 
-    if (values.x) {
-        _axis.x = parseFloat(values.x);
-        updatedValues.x = _axis.x;
+    var sinY = -values.y / G_CONSTANT;
```

Comment:
Why do you say that there is no way to work out rotation from acceleration if it isn't `1g`? I agree that when specifying rotations it is reasonable to assume the device is approximately stationary and thus the acceleration is `1g`, but when specifying acceleration why not compute the magnitude from the given values and work with that? I think the rest of the maths should follow the same, the only special case is when acceleration is 0 in which case it could have any rotation,

That said, it might be a bit confusing to work with, and there are subtleties like `acceleration = 0` means "freefall" rather than "stationary".